Add API Endpoint Test Cases and Mock Service Implementation

Isuranga-2001 opened the pull request:
## Purpose
This pulls request adds test scenarios (with negative testcases) for multiple API endpoints, using a mock service to emulate responses during testing. The main objective is to improve test coverage and verify that the API performs consistently across various situations.

## Examples

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

Isuranga-2001 committed:
add: test.bal file for writing testcases

Isuranga-2001 committed:
add: testcases for basic endpoints (create, get all, get by ID, archive))

Isuranga-2001 committed:
add: testcases for batch endpoints (create, archive and get)

Isuranga-2001 committed:
add: testcase for update existing call

Isuranga-2001 committed:
add: testcase for batch update and search a call by a property

Isuranga-2001 committed:
Use check statement to handle the errors in testcases

Isuranga-2001 committed:
Add negative testcase for GetByID

Isuranga-2001 committed:
Add negative testcases for basic create, search and update endpoints

Isuranga-2001 committed:
Update mock_service.bal: Validate the payload informations in get, delete and update functions

Isuranga-2001 committed:
implement all mock services for testcases

Isuranga-2001 committed:
Merge To: Add API Endpoint Test Cases and Mock Service Implementation

Add API Endpoint Test Cases and Mock Service Implementation

Isuranga-2001 committed:
Changed the environment variables as configurables

Isuranga-2001 committed:
Merge branch 'main' into feat/testcases

Isuranga-2001 committed:
Add Config.toml.template inside of the ballerina/tests

Nuvindu commented on ballerina/tests/Config.toml.template:
let's rename this file as `Config.toml`

Isuranga-2001 replied:
Actually, I used this file as an example for Config.toml. (like ".env.example" file for ".env"). Because the tests directory does not have a markdown file.
![Code](https://github.com/user-attachments/assets/ff0b08ca-da71-4bb7-8fb2-ba8006902fba)

If it is not required, I can remove it.

Nuvindu replied:
You can use the `Config.toml` file instead of `Config.toml.template`. 

However, since `Config.toml` is ignored in `.gitignore`, you won’t be able to commit it. To work around this, temporarily remove `**/Config.toml` from `.gitignore`, commit the file with mock values, and then re-add `**/Config.toml` to `.gitignore`.

Isuranga-2001 replied:
Sure, I will do that.

Isuranga-2001 replied:
I removed the Config.toml.template file. [9828441](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.engagements.calls/pull/9/commits/982844140833f77b739c1911bbe5d7e90d6dc912)

Nuvindu commented on ballerina/tests/test.bal:
Let's use the `Format Document` option to format the code

Isuranga-2001 replied:
Okay

Isuranga-2001 replied:
Done. [3efeddf](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.engagements.calls/pull/9/commits/3efeddf59a6f13a7979d5a564b4d2a240e9177b5)

Isuranga-2001 committed:
Format the test.bal and mock_service.bal using "Format Document" option

Isuranga-2001 committed:
Remove JSON responses from mock_service.bal

Isuranga-2001 committed:
Remove JSON payloads from tests.bal

Isuranga-2001 committed:
Refacor the code by using assertEquals() methods to compare two values instead of assertTrue

Nuvindu commented on ballerina/tests/mock_service.bal:
```diff
@@ -328,12 +216,54 @@ service on new http:Listener(9090) {
+
         http:Response response = new;
-        response.statusCode = http:STATUS_NO_CONTENT;
+        lock {
```
Any reason to use `lock` here? If not, we can remove them

Isuranga-2001 replied:
I think it is required, because without `lock` statement can't access to the shared variables from the `isolated functions`. I used two shared variables within the `tests.bal` to support the test cases.

- `hs_call_id` 
- `hs_batch_call_ids`

Nuvindu replied:
Ack. Better to use `camelCase` for variable names
`hsCallId`, `hsBatchCallIds`

Refer: https://learn-ballerina.github.io/best_practices/naming_conventions.html#variables

Isuranga-2001 replied:
Sure

Isuranga-2001 replied:
Done [268c32e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.engagements.calls/pull/9/commits/268c32eeb1437a4c45c44fb673d1a5b3fe5cca25)

Isuranga-2001 committed:
Remove Config.toml.tempate file

Nuvindu commented on ballerina/tests/mock_service.bal:
```diff
@@ -252,56 +157,48 @@ service on new http:Listener(9090) {
+        int id = 0;
+        lock {
+            // find the id from the call_ids array  
+            foreach var i in call_ids {
+                if i.toString() == callId {
+                    id = i;
+                    break;
+                }
```
```suggestion
            foreach string i in callIds {
                if i == callId {
                    id = i;
                    break;
                }
```

I assumed the `callIds` is a `string[]`. If not try it that way

Isuranga-2001 replied:
Done [89f1ea6](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.engagements.calls/pull/9/commits/89f1ea69ca4bee734a14b785ecc5b0b6d68a8df9)

Isuranga-2001 committed:
Refactor variable names to camel case in test.bal and mock_service.bal

Isuranga-2001 committed:
Refactor call ID handling in mock service

- Changed `callIds` array type from `int[]` to `string[]`.
- Modified `getMaximumCallId` function to convert string IDs to integers for comparison.

Nuvindu commented on ballerina/tests/mock_service.bal:
```diff
@@ -252,56 +157,48 @@ service on new http:Listener(9090) {
+            }
+        }
+        if id == "" {
+            http:Response response = new;
+            response.statusCode = http:STATUS_NOT_FOUND;
+            response.setPayload({"error": "Call ID not found"});
+            return response;
```
we can just return an error here.

```suggestion
            return error("Call ID not found");
```

Isuranga-2001 replied:
Sure, Done [097f935](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.engagements.calls/pull/9/commits/097f9357493e7e04ae90e7c35f555984a18cdee3)

Isuranga-2001 committed:
Refactor error handling in GET and PATCH resources to return errors directly instead of HTTP responses

Nuvindu commented on ballerina/tests/mock_service.bal:
```diff
@@ -438,3 +363,16 @@ service on new http:Listener(9090) {
+        }
+    }
+    return max;
+}
```
```suggestion
}

```

Nuvindu committed:
Add new line at the end of the file
